src/classes/ADDR_Contact_TDTM.cls

```diff
@@ -50,168 +51,195 @@ public class ADDR_Contact_TDTM extends TDTM_Runnable {
-	                // this is handled in the next block, since we do it for both new contacts and updated contacts.
-                }                             
+                manageHouseholdAddrs(contact, i, listOld, dmlWrapper);
+            }
+            
+            // AFTER INSERT
+            if (triggerAction == TDTM_Runnable.Action.AfterInsert) {
+                // for new Contacts that are using address management, create the address object.
+                if (UTIL_CustomSettingsFacade.getSettings().Contacts_Addresses_Enabled__c == true) {
+                    if (!ADDR_Addresses_UTIL.isContactAddressEmpty(contact))
+                        listcontactCreateAddr.add(contact);
+                }
             }
     
-            // BEFORE UPDATE & AFTER UPDATE & BEFORE INSERT && AFTER INSERT
-            // our functionality for dealing with contact address changes requires some trickery!
-            // if the contact doesn't have an address override, we create (or match) a default address,
-            // which requires us to do the work in the AfterUpdate (or else we can't propagate the default address).
-            // but if the contact has an address override, we are supposed to create (or match) a non default address override.
-            // in order to update the contact with the new address override's Id, we have to do it in the BeforeUpdate.
-            if (triggerAction == TDTM_Runnable.Action.BeforeUpdate ||
-               triggerAction == TDTM_Runnable.Action.BeforeInsert ||  
-               triggerAction == TDTM_Runnable.Action.AfterInsert || 
-               triggerAction == TDTM_Runnable.Action.AfterUpdate
-               ) {
-                // note that we detect whether the address change is coming from the address trigger,
-                // in which case we don't want to create a new address.
-                if (ADDR_Addresses_TDTM.hasRunAddrTrigger) {
-                    //UTIL_Debug.debug('DJH**** stop Addr Trigger recursion in ADDR_Contact_TDTM');
-                    continue;
+            // AFTER UPDATE
+            if (triggerAction == TDTM_Runnable.Action.AfterUpdate) {  
+                // if address change is coming from the address trigger we don't want to create a new address   
+                if (!ADDR_Addresses_TDTM.hasRunAddrTrigger) {
+                    if (UTIL_CustomSettingsFacade.getSettings().Contacts_Addresses_Enabled__c == true) {
+                        Contact contactOld = (Contact)listOld[i];
+                        // if the address info has been removed
+                        if(!ADDR_Addresses_UTIL.iscontactAddressEmpty(contactOld) && ADDR_Addresses_UTIL.iscontactAddressEmpty(contact))
+                            listContactsAddrInfoCleared.add(contact);
+                        // if the address changed, remember the Contact we want to add a new address for    
+                        else if (ADDR_Addresses_UTIL.isContactAddressChanged(contact, contactOld)) 
```

Comment:
Nesting too many `IF` statements affects readability and makes maintenance more difficult. I suggest reviewing the design to simplify the flow of this code block.